include/modules/meta/timer_module.hpp

```diff
@@ -34,7 +34,24 @@ namespace modules {
           if (check()) {
             CAST_MOD(Impl)->broadcast();
           }
-          CAST_MOD(Impl)->sleep(m_interval);
+          // wait until next full interval to avoid drifting clocks
+          using clock = chrono::system_clock;
+          using sys_duration_t = clock::time_point::duration;
+
+          auto sys_interval = chrono::duration_cast<sys_duration_t>(m_interval);
+          clock::time_point now = clock::now();
+          sys_duration_t adjusted = sys_interval - (now.time_since_epoch() % sys_interval);
+
+          // The seemingly arbitrary addition of 1ms is due
+          // to the fact that if we wait the exact time our
+          // thread will be woken just a tiny bit prematurely
+          // and therefore the wrong time will be displayed.
+          // Optimally we would round to the next smaller
+          // increment of m_interval (e.g. ms), but the required
+          // functions are c++ 17 only.
+          // For this simple use case, where m_interval is always
+          // in seconds a simple 'magic' constant suffices.
+          CAST_MOD(Impl)->sleep_until(now + adjusted + 1ms);
```

Comment:
Even with the 1ms added, I experience my date module displaying XX:59 with an interval of 60s. I think we can get away with adding 500ms here.

Do we know why the condition_variable wakes up too early? I can understand why it would wake up later, but not earlier.

Reply:
Unfortunately not no.

We initially though it was due to rounding issues but have since ruled that out.
We then suspected possible call reordering (e.g. hoisting both the `std::time_t(nullptr)` call and the `chrono::system_clock::now()` to the beginning of the `runner()` function to maximize cache efficiency, but this also proved to be a fruitless effort.
Spurious wake-ups have also been ruled out as the `condition_variable::wait_until`, which is used internally, does return that the timeout was reached.
It seems like this might be a bug in the `stl` or possibly `libc` (as the `stl` uses [`pthread_cond_wait`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_cond_timedwait.html) internally or due to some other reason which we haven't considered.

In any case, I've raised the magic addition to `500ms`.
Please let me know if you have further possible ideas for explaining this strange behaviour.